<html><head><title>GSE111797</title></head>
<body>
<h1>GSE111797</h1>
<img src="png/GSE111797.png" style="float:right;width:600px;height:400px;">
<li>Series_title: Transcriptome of elt-2 for the nematode Caenorhabditis elegans
<li>Series_summary: In the past decade, the paradigm which claimed that invertebrate immune systems lack specificity has been reconsidered. Accumulating evidence supports that invertebrate immune systems are able to mount specific responses to the pathogen species-, and even to the pathogen strain-level. However, the underlying molecular mechanisms behind invertebrate immune specificity remain mostly unknown. Studying the molecular basis of invertebrate immune specificity in a genetically tractable model, such as the nematode Caenorhabditis elegans, has the potential to reveal insights into the immune systems of other metazoans, including humans. We chose to study the mechanisms of specific immune responses of the worm to two different pathogenic strains of the Gram-positive bacterium Bacillus thuringiensis (MYBY18247 and MYBT18679), because there is phenotypic evidence of specific genotype-genotype interactions between this host-pathogen pair. We did an initial RNA-Seq experiment upon pathogen exposure and found that 9% of the differentially expressed genes change their expression in different ways when comparing the two pathogen strains. Through promoter region motif enrichment analysis, we found the GATA transcription factor ELT-2 is responsible for the pathogen strain-specific transcriptomic response. Upon elt-2 knockdown worms exposed to MYBT18679 display lower survival rate coupled with higher intestinal damage than non-infected controls. Additionally, by performing further genetic analysis using gene knockdown and knockout, we found that the p38 MAPK pathway acts likely in parallel to elt-2 and the transcription factor skn-1 cooperates with elt-2 to promote resistance to MYBT18679. On the other hand, elt-2 knockdown leads to a substantially higher survival rate, together with lower intestinal tissue damage compared to control worms, upon exposure to MYBT18247, another pathogenic Bacillus thuringiensis strain. The MYBT18247 pathogen load of elt-2(RNAi) worms compared to control worms remained unchanged, suggesting the elt-2 negatively regulates tolerance towards MYBT18247. We found that tolerance to MYBT18247 was positively regulated by the transcription factors: FOXO daf-16, bZip zip-2, nhr-99 and nhr-193. To identify elt-2 negatively-regulated downstream targets that could promote tolerance to MYBT18247, we performed a second RNA-Seq experiment, this time including elt-2(RNAi) worms exposed to both pathogenic strains. We found four genes negatively regulated by elt-2: cdr-2, poml-3, dhs-30 and tre-3, with putative function in detoxification and lipid metabolism, which can mediate tolerance to MYBT18247. We conclude that ELT-2 coordinates strain-specific immune responses in this invertebrate host and promotes resistance upon exposure to MYBT18679, while it negatively regulates tolerance to MYBT18247. The response is likely to be specific to the crystal pore-forming toxins produced by this pathogen.
<li>Series_overall_design: mRNA profiles of wild type (WT) and elt-2 (RNAi) C.elegans exposed to either Bacillus thuringiensis B-18247, B-18679 and B-18407 at 6h or 12h were generated by deep sequencing, in triplicate, using Illumina HiSeq2000.
<li>Series_type: Expression profiling by high throughput sequencing
<dl><dt>C1</dt>
<dd><b>GSM3039878</b> Empty_vector_BT679_6h_rep2 | whole body</dd>
<dd><b>GSM3039896</b> Empty_vector_BT247_12h_rep2 | whole body</dd>
<dd><b>GSM3039897</b> Empty_vector_BT247_12h_rep3 | whole body</dd>
<dd><b>GSM3039890</b> Empty_vector_BT247_6h_rep2 | whole body</dd>
<dd><b>GSM3039901</b> Empty_vector_BT407_6h_rep1 | whole body</dd>
<dd><b>GSM3039902</b> Empty_vector_BT407_6h_rep2 | whole body</dd>
<dd><b>GSM3039903</b> Empty_vector_BT407_6h_rep3 | whole body</dd>
<dd><b>GSM3039883</b> Empty_vector_BT679_12h_rep1 | whole body</dd>
<dd><b>GSM3039885</b> Empty_vector_BT679_12h_rep3 | whole body</dd>
<dd><b>GSM3039884</b> Empty_vector_BT679_12h_rep2 | whole body</dd>
<dd><b>GSM3039895</b> Empty_vector_BT247_12h_rep1 | whole body</dd>
<dd><b>GSM3039879</b> Empty_vector_BT679_6h_rep3 | whole body</dd>
<dd><b>GSM3039908</b> Empty_vector_BT407_12h_rep2 | whole body</dd>
<dd><b>GSM3039889</b> Empty_vector_BT247_6h_rep1 | whole body</dd>
<dd><b>GSM3039891</b> Empty_vector_BT247_6h_rep3 | whole body</dd>
<dd><b>GSM3039907</b> Empty_vector_BT407_12h_rep1 | whole body</dd>
<dd><b>GSM3039877</b> Empty_vector_BT679_6h_rep1 | whole body</dd>
<dd><b>GSM3039909</b> Empty_vector_BT407_12h_rep3 | whole body</dd>
</dl>
<dl><dt>C2</dt>
<dd><b>GSM3039904</b> elt-2_RNAi_BT407_12h_rep1 | whole body</dd>
<dd><b>GSM3039880</b> elt-2_RNAi_BT679_12h_rep1 | whole body</dd>
<dd><b>GSM3039888</b> elt-2_RNAi_BT247_6h_rep3 | whole body</dd>
<dd><b>GSM3039886</b> elt-2_RNAi_BT247_6h_rep1 | whole body</dd>
<dd><b>GSM3039892</b> elt-2_RNAi_BT247_12h_rep1 | whole body</dd>
<dd><b>GSM3039882</b> elt-2_RNAi_BT679_12h_rep3 | whole body</dd>
<dd><b>GSM3039906</b> elt-2_RNAi_BT407_12h_rep3 | whole body</dd>
<dd><b>GSM3039898</b> elt-2_RNAi_BT407_6h_rep1 | whole body</dd>
<dd><b>GSM3039875</b> elt-2_RNAi_BT679_6h_rep2 | whole body</dd>
<dd><b>GSM3039905</b> elt-2_RNAi_BT407_12h_rep2 | whole body</dd>
<dd><b>GSM3039893</b> elt-2_RNAi_BT247_12h_rep2 | whole body</dd>
<dd><b>GSM3039881</b> elt-2_RNAi_BT679_12h_rep2 | whole body</dd>
<dd><b>GSM3039900</b> elt-2_RNAi_BT407_6h_rep3 | whole body</dd>
<dd><b>GSM3039894</b> elt-2_RNAi_BT247_12h_rep3 | whole body</dd>
<dd><b>GSM3039874</b> elt-2_RNAi_BT679_6h_rep1 | whole body</dd>
<dd><b>GSM3039876</b> elt-2_RNAi_BT679_6h_rep3 | whole body</dd>
<dd><b>GSM3039887</b> elt-2_RNAi_BT247_6h_rep2 | whole body</dd>
<dd><b>GSM3039899</b> elt-2_RNAi_BT407_6h_rep2 | whole body</dd>
</dl>
</body>
</html>
