<html><head><title>GSE147401</title></head>
<body>
<h1>GSE147401</h1>
<img src="png/GSE147401.png" style="float:right;width:600px;height:400px;">
<li>Series_title: The zinc-finger protein OEF-1 stabilizes patterns of activating and silencing modifications in the C. elegans germ line
<li>Series_summary: Sex-determining chromosomes exhibit complex global regulation in germ cells, in part due to the absence of dosage compensatory mechanisms found in the soma. In the XX hermaphrodite germ line of C. elegans, the X chromosome houses few germline-expressed genes [1] and is poorly expressed, though not completely silent, through most of germ cell development [2]. A chromatin-modifying pathway consisting of the histone methyltransferase (HMT) MES-4 and the Polycomb Repressive Complex 2 (PRC2) orthologs MES-2/3/6 contributes to X silencing in C. elegans germ cells [3-5]. MES-4 promotes H3K36me3 accumulation on autosomes, which leads to concentration of H3K27me3 on the X by MES-2/3/6 [5]. Loss of MES activity results in inappropriate activation of X-linked genes and second-generation sterility [5, 6]. The two marks occupy mutually exclusive domains of the genome [5, 7], leading to a model in which the presence of one modification prevents the accumulation of the other at specific loci. However, further details about how this repulsion mechanism between H3K27me3 and H3K36me3 is established remain mysterious. Here we implicate the zinc-finger protein OEF-1 in counterbalancing the accumulation of H3K27me3 throughout the genome, particularly on the X chromosome. Strikingly OEF-1, like MES-4, is localized to sites of active gene expression in the germ line, and promotes H3K36me3 levels in mes-4 mutants. Our data thus identify OEF-1 as a novel modulator of the fundamental relationship between gene activation and gene silencing in the C. elegans germ line.
<li>Series_overall_design: This data submission includes ChIP-seq samples in C. elegans performed in duplicate for two histone modifications on four genotypes, with one input sample per replicate. In addition, we are submitting a bigwig ile for each dataset. We are also submitting six RNA-seq datasets representing triplicate analysis of wild type and mutant (oef-1(vr25)) dissected gonads from C. elegans.
<li>Series_type: Expression profiling by high throughput sequencing #-# Genome binding/occupancy profiling by high throughput sequencing
<dl><dt>C1</dt>
<dd><b>GSM4430059</b> oef-1(vr25) RNA-seq rep 3 | Dissected adult gonads | Dissected gonads | oef-1(vr25)</dd>
<dd><b>GSM4430058</b> oef-1(vr25) RNA-seq rep 2 | Dissected adult gonads | Dissected gonads | oef-1(vr25)</dd>
<dd><b>GSM4430057</b> oef-1(vr25) RNA-seq rep 1 | Dissected adult gonads | Dissected gonads | oef-1(vr25)</dd>
</dl>
<dl><dt>C2</dt>
<dd><b>GSM4430054</b> N2 RNA-seq rep 1 | Dissected adult gonads | Dissected gonads | wild-type</dd>
<dd><b>GSM4430055</b> N2 RNA-seq rep 2 | Dissected adult gonads | Dissected gonads | wild-type</dd>
<dd><b>GSM4430056</b> N2 RNA-seq rep 3 | Dissected adult gonads | Dissected gonads | wild-type</dd>
</dl>
</body>
</html>
