<html><head><title>GSE165793</title></head>
<body>
<h1>GSE165793</h1>
<img src="png/GSE165793.png" style="float:right;width:600px;height:400px;">
<li>Series_title: An optimized ribodepletion approach for C. elegans RNA-sequencing libraries
<li>Series_summary: Advances over the past decade have allowed for increasingly fine-grained labeling and isolation of rare cell samples for transcriptomic analysis, providing new insights into cell function and gene regulation. These samples often contain very little RNA for sequencing, and so have required new techniques to capture and amplify transcripts of interest. However, as new tools are developed, they are often optimized for mammalian samples. Thus, it is unclear for invertebrate samples what the best practices are for library preparation, and how differences in library preparation approaches affect final results. Here we compared two commercially available techniques: a commonly used polyA selection approach, and a newly developed ribodepletion approach, for which we designed a unique C. elegans-specific probe set. We performed a detailed comparison on multiple RNA samples and developed novel analysis methods to compare the results. Our analysis identifies clear strengths and weaknesses for both of these approaches in building libraries from low abundance RNA samples in C. elegans. We find that the polyA approach is superior in both percent reads mapped and rRNA removal, while the ribodepletion approach provides better detection for long and noncoding RNAs along with increased certainty of calling expression for low abundance transcripts. These insights will help guide researchers in designing future experiments in transcriptomics and provide a model for comparative analysis of library-building techniques.
<li>Series_overall_design: Four RNA samples were used to compare library preparation methods: SMARTseq V4 (Takara) and SoLo Ovation (Tecan Genomics). Each sample was split into two, and used as input for both library prep kits so that final libraries were matched between kits.
<li>Series_type: Expression profiling by high throughput sequencing
<dl><dt>C1</dt>
<dd><b>GSM5050077</b> SMARTseq_2 | FACS C. elegans neurons | otIs355 [rab-3(prom1)::2xNLS-TagRFP] IV</dd>
<dd><b>GSM5050078</b> SMARTseq_3 | FACS C. elegans neurons | otIs355 [rab-3(prom1)::2xNLS-TagRFP] IV</dd>
<dd><b>GSM5050079</b> SMARTseq_4 | FACS C. elegans neurons | otIs355 [rab-3(prom1)::2xNLS-TagRFP] IV</dd>
<dd><b>GSM5050076</b> SMARTseq_1 | FACS C. elegans neurons | otIs355 [rab-3(prom1)::2xNLS-TagRFP] IV</dd>
</dl>
<dl><dt>C2</dt>
<dd><b>GSM5050072</b> SoLo_1 | FACS C. elegans neurons | otIs355 [rab-3(prom1)::2xNLS-TagRFP] IV</dd>
<dd><b>GSM5050075</b> SoLo_4 | FACS C. elegans neurons | otIs355 [rab-3(prom1)::2xNLS-TagRFP] IV</dd>
<dd><b>GSM5050074</b> SoLo_3 | FACS C. elegans neurons | otIs355 [rab-3(prom1)::2xNLS-TagRFP] IV</dd>
<dd><b>GSM5050073</b> SoLo_2 | FACS C. elegans neurons | otIs355 [rab-3(prom1)::2xNLS-TagRFP] IV</dd>
</dl>
</body>
</html>
