<html><head><title>GSE184086</title></head>
<body>
<h1>GSE184086</h1>
<img src="png/GSE184086.png" style="float:right;width:600px;height:400px;">
<li>Series_title: Complete transcriptome of bir-2 knockdown C. elegans nematodes with and without exposure to pathogenic bacteria Enterococcus faecalis.
<li>Series_summary: Purpose: The goals of this study are to examine the role of C. elegans gene bir-2 in innate immunity. The method is to compare differences in gene expression of transcriptomes of C. elegans in the presence and absence of a bir-2 knockdown and with and without pathogenic bacteria. #-# Results: We mapped about 4 million sequence paired-end reads of 251 nucleotide length for each of the 16 samples, trimmed the reads with Trimmomatic (v.0.38), mapped to the WBGene C. elegans genome (build WBcel235.99) using HISAT2 (v.2.1.0), assembled and merged transcripts with StringTie. SamTools were used to convert the BAM files. Differential expression was evaluated using DESEQ2 in R. Approximately 4% of the transcripts showed differential expression between the WT and bir-2 knockdown, with a fold change ?2 and p value <0.05. #-# Methods: C. elegans were grown, in quadruplicate, exposed to HT115 bacteria expressing RNAi for bir-2 with and without E. faecalis exposure. Total RNA was purified and Illumina poly-A selection, library construction (KAPA RNA Hyperprep) and deep sequencing, using Illumina HiSeq2000 at UNH Hubbard Genome Center. The sequence reads were passed through FASTQC and Trimmomatic, and were analyzed at the transcript isoform level. #-# Conclusions: Our study represents the first detailed analysis of retinal transcriptomes, with biologic replicates, generated by RNA-seq technology. The optimized data analysis workflows reported here should provide a framework for comparative investigations of expression profiles. Our results show that NGS offers a comprehensive and more accurate quantitative and qualitative evaluation of mRNA content within a cell or tissue. We conclude that RNA-seq based transcriptome characterization would expedite genetic network analyses and permit the dissection of complex biologic functions.
<li>Series_overall_design: Whole animal mRNA profiles of C. elegans nematodes +/- bir-2 knockdown +/- E. faecalis
<li>Series_type: Expression profiling by high throughput sequencing
<dl><dt>C1</dt>
<dd><b>GSM5577345</b> bir-2knockdown untreated rep4 | Whole animal | Whole animal | bir-2knockdown</dd>
<dd><b>GSM5577347</b> bir-2knockdown pathogen rep2 | Whole animal | Whole animal | bir-2knockdown</dd>
<dd><b>GSM5577346</b> bir-2knockdown pathogen rep1 | Whole animal | Whole animal | bir-2knockdown</dd>
<dd><b>GSM5577348</b> bir-2knockdown pathogen rep3 | Whole animal | Whole animal | bir-2knockdown</dd>
<dd><b>GSM5577343</b> bir-2knockdown untreated rep2 | Whole animal | Whole animal | bir-2knockdown</dd>
<dd><b>GSM5577349</b> bir-2knockdown pathogen rep4 | Whole animal | Whole animal | bir-2knockdown</dd>
<dd><b>GSM5577344</b> bir-2knockdown untreated rep3 | Whole animal | Whole animal | bir-2knockdown</dd>
<dd><b>GSM5577342</b> bir-2knockdown untreated rep1 | Whole animal | Whole animal | bir-2knockdown</dd>
</dl>
<dl><dt>C2</dt>
<dd><b>GSM5577339</b> untreated rep2 | Whole animal | Whole animal | wild type</dd>
<dd><b>GSM5577341</b> untreated rep4 | Whole animal | Whole animal | wild type</dd>
<dd><b>GSM5577338</b> untreated rep1 | Whole animal | Whole animal | wild type</dd>
<dd><b>GSM5577340</b> untreated rep3 | Whole animal | Whole animal | wild type</dd>
</dl>
<dl><dt>C3</dt>
<dd><b>GSM5577351</b> pathogen rep2 | Whole animal | Whole animal | wild type</dd>
<dd><b>GSM5577350</b> pathogen rep1 | Whole animal | Whole animal | wild type</dd>
<dd><b>GSM5577353</b> pathogen rep4 | Whole animal | Whole animal | wild type</dd>
<dd><b>GSM5577352</b> pathogen rep3 | Whole animal | Whole animal | wild type</dd>
</dl>
</body>
</html>
