<html><head><title>GSE223801</title></head>
<body>
<h1>GSE223801</h1>
<img src="png/GSE223801.png" style="float:right;width:600px;height:400px;">
<li>Series_title: DOT-1.1 (DOT1L) deficiency in C. elegans leads to small RNA-dependent gene activation [mRNA, embryos] #-# DOT-1.1 (DOT1L) deficiency in C. elegans leads to small RNA-dependent gene activation
<li>Series_summary: Methylation of histone H3 at lysine 79 (H3K79) is conserved from yeast to humans and is accomplished by Dot1 (disruptor of telomeric silencing-1) methyltransferases. The C. elegans enzyme DOT-1.1 and its interacting partners are similar to the mammalian DOT1L (Dot1-like) complex. The C. elegans DOT-1.1 complex has been functionally connected to RNA interference. Specifically, we have previously shown that embryonic and larval lethality of dot-1.1 mutant worms deficient in H3K79 methylation was suppressed by mutations in the RNAi pathway genes responsible for generation (rde-4) and function (rde-1) of primary small interfering RNAs (siRNAs). This suggests that dot-1.1 mutant lethality is dependent on the enhanced production of some siRNAs. We have also found that this lethality is suppressed by a loss-of-function of CED-3, a conserved apoptotic protease. Here, we describe a comparison of gene expression and primary siRNA production changes between control and dot-1.1 deletion mutant embryos. We found that elevated antisense siRNA production occurred more often at upregulated than downregulated genes. Importantly, gene expression changes were dependent on RDE-4 in both instances. Moreover, the upregulated group, which is potentially activated by ectopic siRNAs, was enriched in protease-coding genes. Our findings are consistent with a model where in the absence of H3K79 methylation there is a small RNA-dependent activation of protease genes, which leads to embryonic and larval lethality. DOT1 enzymesâ€™ conservation suggests that the interplay between H3K79 methylation and small RNA pathways may exist in higher organisms. #-# This SuperSeries is composed of the SubSeries listed below.
<li>Series_overall_design: C. elegans dot-1.1(knu339); ced-3(n1286) mutants and ced-3(n1286) mutant controls (3 replicates each) were age-synchronized and their eggs were harvested via bleaching. RNA was extracted and mRNA and lncRNA sequencing was performed to find differences in gene expression. #-# Refer to individual Series
<li>Series_type: Expression profiling by high throughput sequencing #-# Expression profiling by high throughput sequencing #-# Non-coding RNA profiling by high throughput sequencing
<dl><dt>C1</dt>
<dd><b>GSM6996047</b> dot-1.1(knu339); ced-3(n1286) embryos biol rep 1 | dot-1.1(knu339); ced-3(n1286) mutant embryos | embryos | dot-1.1(knu339); ced-3(n1286)</dd>
<dd><b>GSM6996049</b> dot-1.1(knu339); ced-3(n1286) embryos biol rep 3 | dot-1.1(knu339); ced-3(n1286) mutant embryos | embryos | dot-1.1(knu339); ced-3(n1286)</dd>
<dd><b>GSM6996048</b> dot-1.1(knu339); ced-3(n1286) embryos biol rep 2 | dot-1.1(knu339); ced-3(n1286) mutant embryos | embryos | dot-1.1(knu339); ced-3(n1286)</dd>
</dl>
<dl><dt>C2</dt>
<dd><b>GSM6996051</b> ced-3(n1286) embryos biol rep 2 | ced-3(n1286) mutant embryos | embryos | ced-3(n1286)</dd>
<dd><b>GSM6996052</b> ced-3(n1286) embryos biol rep 3 | ced-3(n1286) mutant embryos | embryos | ced-3(n1286)</dd>
<dd><b>GSM6996050</b> ced-3(n1286) embryos biol rep 1 | ced-3(n1286) mutant embryos | embryos | ced-3(n1286)</dd>
</dl>
</body>
</html>
