<html><head><title>GSE261565</title></head>
<body>
<h1>GSE261565</h1>
<img src="png/GSE261565.png" style="float:right;width:600px;height:400px;">
<li>Series_title: Effects of HOE-1 (RNase Z/ELAC2) nuclear hyperactivation on gene expression in Caenorhabditis elegans.
<li>Series_summary: We previously reported that the RNA endonculease HOE-1 (RNaseZ/ELAC2) is necessary and sufficient for activation of the mitochondrial unfolded protein response (UPRmt) in C. elegans (PMID: 35451962). To comprehensively assess the broader cellular consequences of increased nuclear activity of HOE-1, we conducted RNA sequencing on day 2 adult wildtype animals vs animals in which endogenous HOE-1 is elevated in the nucleus via perturbation of its nuclear export signal (hoe-1(ΔNES)). Additionally, wildtype and hoe-1(ΔNES) animals were grown on both control and atfs-1 RNAi (ATFS-1 is the central transcription factor for UPRmt) to determine gene expression that is atfs-1 dependent. This analysis revealed that differential gene expression in hoe-1(ΔNES) animals exhibits a strong mitochondrial signature. Notably, gene ontology (GO) analysis of significantly upregulated genes showed a significant overrepresentation of mitochondrial associated genes. Moreover, knockdown of atfs-1 strongly compromised the differentially expressed gene profile. Together these data suggest that increased nuclear activity of HOE-1 preferential impacts expression of mitochondrial genes.
<li>Series_overall_design: hoe-1(ΔNES) animals were generated via CRISPR/Cas9 (validation and characterization can be found in PMID: 35451962). #-# hoe-1(ΔNES) animals were then crossed to a conditional sterility mutant (glp-1(e2144). The wildtype condition submitted for RNAseq were glp-1(e2144) animals with wildtype hoe-1. Animals were grown for 72 hours at a non-permissive temperature (25°C) to confer sterility. This allowed for synchronization of adult populations and assessment of only somatic differential gene expression. #-# Both wildtype and hoe-1(ΔNES) animals were grown on control and atfs-1 RNAi (HT115 bacterial background). #-# Three biological replicates of 800 animals each were collected for each condition: wildtype on control RNAi, wildtype on atfs-1 RNAi, hoe-1(ΔNES) on control RNAi, and hoe-1(ΔNES) on atfs-1 RNAi.
<li>Series_type: Expression profiling by high throughput sequencing
<dl><dt>C1</dt>
<dd><b>GSM8146724</b> glp-1(e2144)_control RNAi_rep2 | whole animal somatic | whole animal somatic | glp-1(e2144)</dd>
<dd><b>GSM8146726</b> hoe-1(mpt67);glp-1(e2144)_control RNAi_rep1 | whole animal somatic | whole animal somatic | hoe-1(mpt67);glp-1(e2144)</dd>
<dd><b>GSM8146725</b> glp-1(e2144)_control RNAi_rep3 | whole animal somatic | whole animal somatic | glp-1(e2144)</dd>
<dd><b>GSM8146728</b> hoe-1(mpt67);glp-1(e2144)_control RNAi_rep3 | whole animal somatic | whole animal somatic | hoe-1(mpt67);glp-1(e2144)</dd>
<dd><b>GSM8146727</b> hoe-1(mpt67);glp-1(e2144)_control RNAi_rep2 | whole animal somatic | whole animal somatic | hoe-1(mpt67);glp-1(e2144)</dd>
<dd><b>GSM8146723</b> glp-1(e2144)_control RNAi_rep1 | whole animal somatic | whole animal somatic | glp-1(e2144)</dd>
</dl>
<dl><dt>C2</dt>
<dd><b>GSM8146734</b> hoe-1(mpt67);glp-1(e2144)_atfs-1 RNAi_rep3 | whole animal somatic | whole animal somatic | hoe-1(mpt67);glp-1(e2144)</dd>
<dd><b>GSM8146732</b> hoe-1(mpt67);glp-1(e2144)_atfs-1 RNAi_rep1 | whole animal somatic | whole animal somatic | hoe-1(mpt67);glp-1(e2144)</dd>
<dd><b>GSM8146733</b> hoe-1(mpt67);glp-1(e2144)_atfs-1 RNAi_rep2 | whole animal somatic | whole animal somatic | hoe-1(mpt67);glp-1(e2144)</dd>
</dl>
<dl><dt>C3</dt>
<dd><b>GSM8146730</b> glp-1(e2144)_atfs-1 RNAi_rep2 | whole animal somatic | whole animal somatic | glp-1(e2144)</dd>
<dd><b>GSM8146729</b> glp-1(e2144)_atfs-1 RNAi_rep1 | whole animal somatic | whole animal somatic | glp-1(e2144)</dd>
<dd><b>GSM8146731</b> glp-1(e2144)_atfs-1 RNAi_rep3 | whole animal somatic | whole animal somatic | glp-1(e2144)</dd>
</dl>
</body>
</html>
