<html><head><title>GSE246966</title></head>
<body>
<h1>GSE246966</h1>
<img src="png/GSE246966.png" style="float:right;width:600px;height:400px;">
<li>Series_title: Microbiome remodeling through bacterial competition and host behavior enables rapid adaptation to environmental toxins
<li>Series_summary: Human activity is altering the environment at a rapid pace, challenging the adaptive capacities of genetic variation within animal populations. Animals also harbor extensive gut microbiomes, which play diverse roles in host health and fitness and may help expanding host capabilities. The unprecedented scale of human usage of xenobiotics and contamination with environmental toxins describes one challenge against which bacteria with their immense biochemical diversity are particularly suited to offer solutions. To explore the paths leading to bacteria-assisted rapid adaptation, we used Caenorhabditis elegans harboring a defined microbiome, and the antibiotic neomycin as a model toxin, harmful for the worm host and neutralized to different extents by microbiome members. Worms exposed to neomycin showed delayed development and decreased survival but were protected when colonized by neomycin-resistant members of the microbiome. Through a combination of 16S gene sequencing, counting of live bacteria and behavioral assays we identified two distinct mechanisms that facilitated adaptation: gut enrichment for a neomycin-modifying strain driven by altered bacterial competition; and host avoidance behavior, which depended on the stress-activated KGB-1/JNK and enabled preference of neomycin-protective bacteria. The straightforwardness of these mechanisms suggests that bacteria-assisted host adaptation may be more common than currently appreciated, protecting animals from novel stressors. However, gut remodeling may also cause dysbiosis, and additional experiments identified fitness trade-offs including increased susceptibility to infection as well as metabolic remodeling. Extending these results to other toxins suggests yet unaccounted-for microbiome-dependent long-term consequences of toxin exposure.
<li>Series_overall_design: To investigate how neomycin affects C. elegans, worms raised on E. coli or a complex community (CeMbio) were subjected to 0.25mg/ml neomycin for 8 hours. Comparative gene expression profiling was performed using the RNA-seq data of worms with or without exposure to neomycin with respect to each bacterial community.
<li>Series_type: Expression profiling by high throughput sequencing
<dl><dt>C1</dt>
<dd><b>GSM7883272</b> N2_CeMbio, no antibiotic,3 | metagenomics | metagenomics | WT</dd>
<dd><b>GSM7883264</b> N2,OP50,no antibiotic,1 | metagenomics | metagenomics | WT</dd>
<dd><b>GSM7883271</b> N2_CeMbio, no antibiotic,2 | metagenomics | metagenomics | WT</dd>
<dd><b>GSM7883265</b> N2,OP50,no antibiotic,2 | metagenomics | metagenomics | WT</dd>
<dd><b>GSM7883266</b> N2,OP50,no antibiotic,3 | metagenomics | metagenomics | WT</dd>
<dd><b>GSM7883270</b> N2_CeMbio, no antibiotic,1 | metagenomics | metagenomics | WT</dd>
</dl>
<dl><dt>C2</dt>
<dd><b>GSM7883274</b> N2, CeMbio, neomycin, 2 | metagenomics | metagenomics | WT</dd>
<dd><b>GSM7883269</b> N2,OP50,neomycin,3 | metagenomics | metagenomics | WT</dd>
<dd><b>GSM7883268</b> N2,OP50,neomycin,2 | metagenomics | metagenomics | WT</dd>
<dd><b>GSM7883267</b> N2,OP50,neomycin,1 | metagenomics | metagenomics | WT</dd>
<dd><b>GSM7883273</b> N2, CeMbio, neomycin, 1 | metagenomics | metagenomics | WT</dd>
<dd><b>GSM7883275</b> N2, CeMbio, neomycin, 3 | metagenomics | metagenomics | WT</dd>
</dl>
</body>
</html>
