<html><head><title>GSE228002</title></head>
<body>
<h1>GSE228002</h1>
<img src="png/GSE228002.png" style="float:right;width:600px;height:400px;">
<li>Series_title: Foreign RNA spike-ins enable accurate allele-specific expression analysis at scale [I] #-# Foreign RNA spike-ins enable accurate allele-specific expression analysis at scale
<li>Series_summary: Analysis of allele-specific expression is strongly affected by the technical noise present in RNA-seq experiments. Previously, we showed that technical replicates can be used for precise estimates of this noise, and we provided a tool for correction of technical noise in allele-specific expression analysis. This approach is very accurate but costly due to the need for two or more replicates of each library. Here, we develop a spike-in approach that is highly accurate at only a small fraction of the cost. #-# We show that a distinct RNA added as a spike-in before library preparation reflects technical noise of the whole library and can be used in large batches of samples. We experimentally demonstrate the effectiveness of this approach using combinations of RNA from species distinguishable by alignment, namely, mouse, human, and C.elegans. Our new approach, controlFreq, enables highly accurate and computationally efficient analysis of allele-specific expression in (and between) arbitrarily large studies at an overall cost increase of âˆ¼ 5%. #-# This SuperSeries is composed of the SubSeries listed below.
<li>Series_overall_design: Comparative allele-specific expression analysis of RNA-seq data with foreign RNA spike-ins (mixes of human, mouse and C.elegans). #-# Refer to individual Series #-# grantee -- *Alexander Gimelbrant* #-# grant # -- *5 R01 CA262577-02* #-# grant title -- *Elucidating prostate cancer risk mechanisms through #-# large-scale cistrome wide association studies* #-# name of the funding source -- *NIH*
<li>Series_type: Expression profiling by high throughput sequencing #-# Expression profiling by high throughput sequencing
<dl><dt>C1</dt>
<dd><b>GSM7112261</b> W1_repl2 | N2 and Hawaii strains | WT</dd>
<dd><b>GSM7112260</b> W1_repl1 | N2 and Hawaii strains | WT</dd>
<dd><b>GSM7112262</b> W1_repl3 | N2 and Hawaii strains | WT</dd>
</dl>
<dl><dt>C2</dt>
<dd><b>GSM7112241</b> M1W1_50vs50_pRNA_repl3 | v-Abl pro-B clonal cell line Abl.1 / N2 and Hawaii strains | WT</dd>
<dd><b>GSM7112236</b> H1W1_85vs15_pRNA_repl1 | clonal pro-B cell line GM12878.DF1 / N2 and Hawaii strains | WT</dd>
<dd><b>GSM7112226</b> M1W1_27vs73_pRNA_repl3 | v-Abl pro-B clonal cell line Abl.1 / N2 and Hawaii strains | WT</dd>
<dd><b>GSM7112231</b> H1W1_65vs35_pRNA_repl2 | clonal pro-B cell line GM12878.DF1 / N2 and Hawaii strains | WT</dd>
<dd><b>GSM7112224</b> M1W1_27vs73_pRNA_repl1 | v-Abl pro-B clonal cell line Abl.1 / N2 and Hawaii strains | WT</dd>
<dd><b>GSM7112239</b> M1W1_50vs50_pRNA_repl1 | v-Abl pro-B clonal cell line Abl.1 / N2 and Hawaii strains | WT</dd>
<dd><b>GSM7112237</b> H1W1_85vs15_pRNA_repl2 | clonal pro-B cell line GM12878.DF1 / N2 and Hawaii strains | WT</dd>
<dd><b>GSM7112238</b> H1W1_85vs15_pRNA_repl3 | clonal pro-B cell line GM12878.DF1 / N2 and Hawaii strains | WT</dd>
<dd><b>GSM7112232</b> H1W1_65vs35_pRNA_repl3 | clonal pro-B cell line GM12878.DF1 / N2 and Hawaii strains | WT</dd>
<dd><b>GSM7112225</b> M1W1_27vs73_pRNA_repl2 | v-Abl pro-B clonal cell line Abl.1 / N2 and Hawaii strains | WT</dd>
<dd><b>GSM7112240</b> M1W1_50vs50_pRNA_repl2 | v-Abl pro-B clonal cell line Abl.1 / N2 and Hawaii strains | WT</dd>
<dd><b>GSM7112230</b> H1W1_65vs35_pRNA_repl1 | clonal pro-B cell line GM12878.DF1 / N2 and Hawaii strains | WT</dd>
</dl>
</body>
</html>
